crates/parser/tests/parser_tests.rs

```diff
@@ -1326,6 +1330,55 @@ x %= 4";
         }
     }
 
+    mod variable_declaration {
+        use super::*;
+
+        #[test]
+        fn single() {
+            let source = "let a = 1";
+
+            check_ast(
+                source,
+                &[
+                    id(0),
+                    SmallInt(1),
+                    Assign {
+                        target: 0.into(),
+                        expression: 1.into(),
+                    },
+                    MainBlock {
+                        body: vec![2.into()],
+                        local_count: 1,
+                    },
+                ],
+                Some(&[Constant::Str("a")]),
+            )
+        }
+
+        #[test]
+        fn single_with_type_hint() {
+            let source = "let a: Number = 1";
+
+            check_ast(
+                source,
+                &[
+                    id(0),
+                    type_hint(1, &[]),
```

Comment:
The type hint here will need to be pushed into the AST before the ID, so that the ID can refer to the type hint by AST index. In the parser, `consume_type_hint` should push the node into the AST and then return the hint's `AstIndex`, which can then be used when pushing the `Id` node.